File: fsharp/grep/Grep.fs
module Grep
open System.IO

let grep files flagArguments pattern = 
    let arg x = List.contains x flagArguments
    let lower (s:string) = if arg "-i" then s.ToLower() else s
    let matcher (_,text) = 
        let pattern = lower pattern
        let text = lower text
        let valid = if arg "-x" then text = pattern else text.Contains pattern
        if arg "-v" then not valid else valid
    let format file (num,text) = 
        match flagArguments with
        | _ when arg "-l" -> file
        | _ ->
            [
                (if List.length files > 1 then file + ":" else "");
                (if arg "-n" then num.ToString() + ":" else "");
                text
            ] |> System.String.Concat
    let rec innerGrep files flagArguments pattern = 
        seq {
            let file = List.head files
            let lines = File.ReadAllLines file |> Seq.mapi (fun i l -> (i+1,l))
            yield! lines |> Seq.filter matcher |> Seq.map (format file)
            match List.tail files with
            | [] -> ()
            | rest -> yield! innerGrep rest flagArguments pattern
        } |> Seq.distinct |> Seq.toList
    innerGrep files flagArguments pattern

File: csharp/grep/Grep.cs
using System;
using System.IO;
using System.Linq;

public static class Grep
{
    public static string Match(string pattern, string flags, string[] files)
    {
        Func<(string file, int index, string text), bool> matcher = line => line.text.Contains(pattern);
        Func<(string file, int index, string text), string> printer = line => line.text;

        bool flag(string token) => flags.Contains(token);

        if(flag("-x"))
            matcher = line => line.text == pattern;
        if(flag("-i"))
        {
            pattern = pattern.ToLower();
            var innerMatcher = matcher;
            matcher = line => innerMatcher((line.file, line.index, line.text.ToLower()));
        }
        if(flag("-v"))
        {
            var innerMatcher = matcher;
            matcher = line => !innerMatcher(line);
        }   

        if(flag("-n"))
            printer = line => $"{line.index + 1}:{line.text}";
        if(flag("-l"))
            printer = line => line.file;
        else if(files.Length > 1)
        {
            var innerPrinter = printer;
            printer = line => $"{line.file}:{innerPrinter(line)}";
        }

        var results = files
            .SelectMany(fileName => 
                File.ReadAllLines(fileName).Select((line, index) => 
                    (fileName, index, line)))
            .Where(matcher).Select(printer).Distinct();

        return string.Join("\n", results);
    }
}